analyzer/runtime/evm/erc721.go

```diff
@@ -62,6 +84,57 @@ func evmDownloadTokenERC721(ctx context.Context, logger *log.Logger, source node
 	return &tokenData, nil
 }
 
+func evmDownloadNFTERC721(ctx context.Context, logger *log.Logger, source nodeapi.RuntimeApiLite, ipfsClient ipfsc.Client, round uint64, tokenEthAddr []byte, id *big.Int) (*EVMNFTData, error) {
+	var nftData EVMNFTData
+	supportsMetadata, err := detectInterface(ctx, logger, source, round, tokenEthAddr, ERC721MetadataInterfaceID)
+	if err != nil {
+		return nil, fmt.Errorf("checking ERC721Metadata interface: %w", err)
+	}
+	if supportsMetadata {
+		if err = evmCallWithABI(ctx, source, round, tokenEthAddr, evmabi.ERC721Metadata, &nftData.MetadataURI, "tokenURI", id); err != nil {
+			if !errors.Is(err, EVMDeterministicError{}) {
+				return nil, fmt.Errorf("calling tokenURI: %w", err)
+			}
+			logDeterministicError(logger, round, tokenEthAddr, "ERC721Metadata", "tokenURI", err,
+				"nft_id", id,
+			)
+		}
+		if nftData.MetadataURI != "" {
+			logger.Info("downloading metadata",
+				"token_eth_addr", hex.EncodeToString(tokenEthAddr),
+				"token_id", id,
+				"uri", nftData.MetadataURI,
+			)
+			nftData.MetadataAccessed = time.Now()
+			rc, err1 := drawerfullofprotocols.Get(ctx, ipfsClient, nftData.MetadataURI)
+			if err1 != nil {
+				// todo: retry on some errors?
```

Comment:
big implications here, from the possibility that the metadata request goes to something other than our own node

1. the server could be slow. even maliciously slow. there's a 30s timeout, but like that means someone could make every batch take 30 seconds
2. some jobs can now fail without it being a horrible "our node is down and nothing else will ever work" situation, as was the case before. we'll want a way to continue with other work
3. some failures probably will be intermittent. after all, these requests will go over the internet, right? we'll want to hang on to a job so we can retry it later
4. oh and "later" will best be one of those backoff systems
5. all that, while we shouldn't assume that the list of all NFTs fits in memory

Reply:
I'm trying to foresee what the architecture will be like, longer term. I expect there will be several async phases:
- Token id -> token metadata URL (by issuing an EVM query to the runtime).
- Metadata fetching (URL -> JSON) 
- Metadata analysis. This is almost trivial; just extracting the relevant field(s) from the metadata JSON.
- Image fetching (URL -> blob)

The reason I think it probably makes sense for each of these to be async from other is that the URL fetching is a) fallible and b) a reusable component.

Phases 2 and 4 are pretty much identical when it comes to requirements: URLs are untrusted, the fetched resource might become stale (I'm referring to non-constant NFTs here), fetching might need to be retried multiple times with backoff. So one possible solution would be to have a "resource fetching" analyzer: Stages 1 and 3 give jobs to it, and tag the job somehow to signify the table into which the result should be output. So the analyzer has the understanding of those tags built in. Or, it could be two separate analyzers, with a largely shared codebase, but each instantiated with a separate `dataInserter` struct (much like `processor` for blockAnalyzers) that knows what to do with the fetched data.

I don't think we need all of the above in this PR (or that it'd be an appropriate amount of new code for one PR, at that :)), but it would probably be helpful to align things with the architecture we want to move towards. For example, if we go with what I outlined above, I think it would make sense to add a minimal `resourceFetchAnalyzer` now. No retries, cache expiration, etc, just a naive one-off fetch.

Depending on how much our thought (mis)align, this might be worth a short doc, actually. I think neither the structure of the NFT analyzer(s) nor the structure of a resource fetcher/cache is clear yet.

Reply:
Also wanted to point out that only phase 1 is EVM specific, which would be nice to reflect in analyzer names. (Just `nft`, no `evm_nft`)

Reply:
oh I forgot to remove the "evm" part

Reply:
okay counter proposal: as this analyzer still has the EVM part, let's leave the "evm" in the name. we can separate out the network download later, along with tackling server errors

Reply:
sgtm for the rename.

For the network download: if I understand the code correctly, any failed downloads currently will just cause the token to stay in the analyzer queue, so if we add a more complex `downloadAnalyzer` later, it will "just work" in that we won't need any complex db migrations to construct a work queue for it. If so, SGTM. Can you create a ticket in that case, please?

Reply:
> any failed downloads currently will just cause the token to stay in the analyzer queue

no, there's more fiddliness to it. right now:

"DONE" means download job is successfully completed, possibly with no or partial info collected; "BREAK" means something really bad is happening and abort the whole batch. currently there's no third option for "just this one has a problem"

- problem checking for ERC721Metadata ->
  - deterministic -> ERC165 says this means it's not supported. (continue as if not supported)
  - nondeterministic -> something's wrong with our access to the network. BREAK
- doesn't support ERC721Metadata -> that's fine, no data from this source. DONE
- supports ERC721Metadata -> (query tokenURI)
  - problem querying tokenURI ->
    - deterministic -> contract's fault. DONE
    - nondeterministic -> something's wrong with our access to the network. BREAK
  - got token URI -> (save token URI and "accessed" date)
    - URI empty or malformed etc. -> contract's fault. DONE
    - URI unsupported scheme -> uh... just give up. DONE
    - URI parseable and supported -> (try to download)
      - just about anything goes wrong -> not necessarily NFT's fault, but also don't want to let this starve the queue. DONE :arrow_left: **this thing**
      - download OK -> (try to parse the info JSON doc)
        - problem parsing JSON -> NFT's fault. DONE
        - JSON OK -> close download stream
          - problem closing download stream -> Go runtime integrity problem?? BREAK
          - close stream OK -> (save name + description + image URL) DONE

so the main thing labeled "this thing" above, that's where currently we treat that as "DONE" and it's arguably bad. this PR has the beginnings of some things that'll help us tell if it's our fault or the NFT's fault, but it's not in use yet. we'll probably have to do a migration to repopulate the queue with jobs that ended up this way. a superset of those jobs, really

but still yeah, filed in #532.

Reply:
:skull: resolving this to focus on active discussions